### PR TITLE
DBZ-6472 MySqlSnapshotChangeEventSource parallel execute createSchemaEventsForTables

### PR DESCRIPTION
**Pull request by harveyyue**
https://issues.redhat.com/browse/DBZ-6472

### CODE REVIEW

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java [hunk 1]
```diff
@@ -333,38 +346,77 @@ protected void readTableStructure(ChangeEventSourceContext sourceContext,
...
+            throws Exception {
+        CompletionService<Map<TableId, String>> completionService = new ExecutorCompletionService<>(executorService);
+        List<TableId> realTablesToRead;
+        if (firstPhase) {
+            realTablesToRead = tablesToRead.stream()
+                    .filter(id -> !delayedSchemaSnapshotTables.contains(id))
+                    .collect(Collectors.toList());
+        }
+        else {
+            realTablesToRead = new ArrayList<>(tablesToRead);
```
Comment (mfvitale) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java:390]: Maybe you can simply this removing the else and move initialization to line 383.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java [hunk 2]
```diff
@@ -333,38 +346,77 @@ protected void readTableStructure(ChangeEventSourceContext sourceContext,
...
+        }
+
+        if (realTablesToRead.size() > 0) {
+            for (TableId tableId : realTablesToRead) {
+                completionService.submit(() -> {
+                    JdbcConnection connection = connectionPool.poll();
+                    assert connection != null;
+                    try {
+                        return connection.queryAndMap("SHOW CREATE TABLE " + quote(tableId), rs -> {
```
Comment (mfvitale) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlSnapshotChangeEventSource.java:399]: WDYT to wrap this into a method to improve readability?

**Comment by mfvitale** [comment]
Can this changes be generic and not MySQL specific?

**Comment by harveyyue** [comment]
> Can this changes be generic and not MySQL specific?

I think it's not generic, since only the MySQL connector use the statement `show create table table_name` to get the schema per table, the parallel reading mechanism will improve this step.

**Comment by harveyyue** [comment]
> @harveyyue I don't think we can do it this way. MySQL by default tries to obtain global read lock to capture schema snapshot. DOn't we end up with serial execution anyway as each of the threads will wait for getting the lock one by one?

Yes, you are right. I didn't consider this lock scenario, maybe it can be used to `none` locking mode.

**Comment by jpechane** [comment]
@harveyyue Yes, in that case it could be separated. Could you please refactor the code to reflect that fact?

**Comment by harveyyue** [comment]
> @harveyyue Yes, in that case it could be separated. Could you please refactor the code to reflect that fact?

Okay

**Comment by harveyyue** [comment]
> @harveyyue I don't think we can do it this way. MySQL by default tries to obtain global read lock to capture schema snapshot. DOn't we end up with serial execution anyway as each of the threads will wait for getting the lock one by one?

Done.

**Comment by jpechane** [comment]
@harveyyue LGTM! Could you please add test(s) to it too?

**Comment by harveyyue** [comment]
> @harveyyue LGTM! Could you please add test(s) to it too?

OK

**Review by vjuranek — APPROVED** [review]

**Comment by jpechane** [comment]
@harveyyue Applied, thanks!

**Pull request merged**